### ISSUE TITLE
Finalizer which creates new garbage can cause infinite loop

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#310]
So, remembering how the finalizer torture test had to take pains to avoid generating new refzero entries, I realized this can be an issue in Ecmascript code as well, and came up with this testcase to illustrate it:

```javascript
function Majin(name, power) {
	this.blowUpTheEarth = function() {
		if (power < 1000000000) { print(name + " isn't powerful enough..."); }
		else { print(name + " blew up the Earth!"); }
	}
}
Duktape.fin(Majin.prototype, function() {
	var buu = new Majin("Majin Buu", 1000000000);
	buu.blowUpTheEarth();
});
new Majin("Vegeta", 9000).blowUpTheEarth();
Duktape.gc();
```

The end result: Buu proceeds to blow up the Earth from now until the end of time. :)

This testcase is admittedly a bit contrived (it's half a step away from being deliberate infinite recursion), but it would easily be possible to cause this by accident if the finalizer needed to create other kindsof objects with finalizers, which themselves have finalizers, etc.

**Comment by fatcerberus** [comment]
I can also generate a segfault from this: minisphere performs a longjmp if you click the game window's Close button and cleans up (this was the easiest way I could find to unconditionally back out of JS execution).  Duktape then crashes because the refzero queue wasn't empty when I attempted to shut down (an assert is triggered if those are enabled).

**Comment by fatcerberus** [comment]
I assume this kind of thing is why objects with finalizers in .NET take two GC passes to free... strange, this never occurred to me before now.

**Comment by svaarala** [comment]
That's more or less a "feature": what else could Duktape do other than finalize the new garbage (with a finalizer) user code created?

**Comment by svaarala** [comment]
I found that crash using torture testing yesterday and there's a fix for it in the #287 work branch.

**Comment by svaarala** [comment]
It would be possible to have a sanity limit for the refzero list processing so that it'd only process, say, 10000 objects and then exit, picking up finalizing objects when it next started.

But that wouldn't really solve the problem. For example, if my finalizer creates *two* finalizable objects, you'll run out of memory regardless of how the finalization works. That is the case at least if you *guarantee* finalizers will be called. If you don't, then it's easy to fix, but will be very awkward for users ("so what are the guarantees for finalization again?" :-)

**Comment by svaarala** [comment]
The current refzero/finalizer behavior is "immediate" so that:

- Refzero may cause other objects to have refcount zero, and that whole cascade is processed immediately
- Finalizers are executed immediately, and if they generate garbage, they're also processed immediately

Relaxing either one would be possible. Some game developers have requested for finalizers to be run separately one by one (some users requested the ability to run finalizers one by one, e.g. one per frame when there's time). But even so, if a finalizer creates more than one new finalizable object, you'll be in trouble unless finalization is skipped.

**Comment by fatcerberus** [comment]
Hm.  It might be interesting to find out how other environments like the .NET CLR or Lua (does Lua have finalizers?) handles this type of thing.  And good point on more than one object being created, I hadn't considered that case.

**Comment by svaarala** [comment]
Incremental GCs typically need two passes over finalizable objects to be sure they're not referenced. If the finalizers then generate more garbage, that gets another two passes etc.

**Comment by fatcerberus** [comment]
I'd guess it still ultimately breaks down if the finalizers themselves create more than one new finalizable object... although I think .NET might not guarantee finalization? Hm.  I definitely don't want to give up guaranteed finalization, my OCD side wouldn't like that very much. ;-)

I suppose it would suffice to mention this as a potential issue in the documentation, i.e. as a rule of thumb don't create new garbage in a finalizer if it can be avoided.

**Comment by svaarala** [comment]
Yeah, this would definitely be a useful thing to mention in finalizer documentation: a finalizer should create, on the average, less than 1 new object with a finalizer :-)

**Comment by svaarala** [comment]
Just to clarify the previous: it's fine for a finalizer to create garbage, but not systematically create garbage with a finalizer. Even that is OK if you know there's not going to be an infinite loop - e.g. a finalizer for object type A can create a temporary object of type B which has a finalizer. But if B then creates a temporary object of type A you'll run into trouble.

**Comment by fatcerberus** [comment]
Yeah, I realize that.  Like I said my testcase above is actually pretty contrived because the object creates more of itself in its finalizer, essentially an infinite loop by a different name. ;-)

**Comment by svaarala** [comment]
Yeah, I was trying to clarify the point to make in the documentation. Also fixed the title because garbage as such is fine :)

**Comment by svaarala** [comment]
By the way, I think the executor timeout should break out such a finalizer loop, but only if the finalizer is an Ecmascript function. If the finalizer is a Duktape/C function there are no executor interrupts to trigger a timeout callback. This is intentional now but it's a "known issue" that a bytecode executor timeout doesn't catch all relevant timeout cases. This is in line with the current assumption that C code is expected to be written carefully (because it can segfault or forever loop trivially anyway) but may not be obvious.

Independent of this it'd be nice to trigger "executor interrupt" like behavior from e.g. call handling to enable both script timeouts *and* debugger operations even when no Ecmascript code is executed in a certain part of the code.

**Comment by fatcerberus** [comment]
I'm fine with no timeouts for C code, especially since Duktape is already designed to fully trust its caller anyway (no NULL check for `ctx` etc.)

**Comment by svaarala** [comment]
There is actually no way *not* to trust C code quite a lot - there's no way for an embedded engine to prevent infinite loops (without interacting with native threading) or memory unsafe behavior. So it's not really a design choice at that level. The implication for sandboxing is just to make it clear what kinds of things would break the sandboxing behavior one is trying to guarantee for Ecmascript code. For example, if you use script timeouts, Duktape/C bindings shouldn't compromise this by doing an infinite loop with some invalid argument or such. Not writing finalizers that potentially cause an infinite finalization loop is a similar compromise.

That NULL check for `ctx` is true, but is more an exception than the rule: mostly all pointers are checked but `ctx` is not because it'd have a measurable performance impact for e.g. stack manipulation. This impact is amplified because Duktape uses the same API quite a lot internally.

**Comment by fatcerberus** [comment]
So I guess there's nothing to fix here... should I close this?

**Comment by svaarala** [comment]
If there's a workable solution this would definitely be worth handling. I hesitate to say "fixing" because it's really broken application code here, and there isn't a clean solution: either one compromises and at some point won't run the finalizer (which may leak native resources or worse) or just keeps on running finalizers until they are exhausted.

There is already a pretty sophisticated approach for object rescue. Maybe that could be tweaked to also handle this case somehow. For example, after a certain number of rounds we'd stop running finalizers and that would then just be documented as a feature to get the heap shut down if there are finalizers bugs. Maybe that's preferable to being stuck in an infinite loop - especially because not *all* finalizer bugs would involve native resource leaks.
